peer/upstream/setup.go

```diff
@@ -113,15 +113,14 @@ loop:
 
 			u.RLock()
 			if u.connected {
-				u.RUnlock()
 				err := u.push(&item)
 				if nil != err {
 					log.Errorf("push: error: %s", err)
 				}
 			} else {
-				u.RUnlock()
 				log.Warn("upstream has not connected")
 			}
+			u.RUnlock()
```

Comment:
`err := u.push(&item)` hold the lock inside. I think for better performance, we just want to lock something we need to get/set and release ASAP. In this case, we only need to lock for getting `u.connected` and we should remove the lock after getting the value.

Reply:
Also `u.push(&item)` is network task, so it would hold the lock in a quite long time, i afraid the performance would be affected.

Reply:
Yeah, that makes sense!!! Tks, I'm gonna close this PR then.